---
level: 0
description: Research Tools for RNA Modeling &amp; Design
---

# RiboKit

[![RiboKit Logo](/assets/ribokit.gif "RiboKit Logo"){: .full}](/assets/ribokit.gif)
{: .center}

Research Tools for RNA Modeling &amp; Design, by the Das Lab at Department of Biochemistry, School of Medicine, Stanford University.

<hr/>
## Multidimensional Chemical Mapping

| Package | Description | Language |
| --- | --- | ---: |
| [**Biers**](/biers) | From MCM (MOHCA) to 3D | `MATLAB` |
| [**HiTRACE**](/hitrace) | The standard tool for MCM analysis of CE | `MATLAB` |
| ~~**LIFFT**~~ <small>(external)<small> | The standard tool for titration analysis | `MATLAB` |
| ~~**Map2D**~~ <small>(external)<small> | The standard tool for MCM analysis of RING-MaP | `Python` |
| ~~**MAPSeeker**~~ <small>(external)<small> | The standard tool for MCM analysis of NGS | `MATLAB` / `C++` |
| [**Primerize**](/primerize) | First stop for assembling templates for RNA synthesis | `Python` |
| [**RDATKit**](/rdatkit) | Utils for anyone working with MCM data | `Python` |
| [**REEFFIT**](/reeffit) <small>(external)<small> | The standard tool for MCM analysis of multi-state RNAs | `Python` |

<hr/>
## Modeling &amp; Design

| Package | Description | Language |
| --- | --- | ---: |
| ~~**DasLab Tools**~~ <small>(external)<small> | Daslab-specific tools | `Python` |
| ~~**ERRASER**~~ <small>(external)<small> | Routine step in expt RNA structural biology | `Python` |
| ~~**EternaBot**~~ <small>(external)<small> | Standard RNA 2D design tool | `Python` |
| ~~**HelixMC**~~ <small>(external)<small> | Modeling optical tweezers experiments | `Python` |
| ~~**RiboLogic**~~ <small>(external)<small> | Standard RNA 2D design tool | `Python` |
| ~~**RNAmake**~~ <small>(external)<small> | "Vienna" for 3D RNA folding/design | `Python` |
| ~~**RNA Tools**~~ <small>(external)<small> | Utils for work on RNA modeling | `Python` |
| ~~**Rosetta RNA _de novo_**~~ <small>(external)<small> | State-of-the-art solver for complex RNA 3D folds | `C++` |
| ~~**Rosetta RNAdesign**~~ <small>(external)<small> | Fast generator of RNA design hypotheses | `C++` |
| ~~**Rosetta Stepwise**~~ <small>(external)<small> | State-of-the-art solver for high-res motif modeling/design | `C++` |
| ~~**RW3D**~~ <small>(external)<small> | Ribokit's tool to kick off Rosetta 3D modeling jobs | `Python` |
| ~~**Stepwise Benchmark**~~ <small>(external)<small> | Benchmark scripts for 3D modeling tools | `Python` |

<hr/>
## Sequence Analysis

| Package | Description | Language |
| --- | --- | ---: |
| ~~**BPPAlign**~~ <small>(external)<small> | Discovery of new rna structure from sequence alignments | `Python` |

<hr/>
## Miscellaneous

| Package | Description | Language |
| --- | --- | ---: |
| [**Pymol DasLab**](/pymol_daslab/) | Utils for anyone working on RNA modeling | `Python` |
| ~~**RMDB API**~~ <small>(external)<small> | "PDB" for MCM data | |

<hr/>
## Bonus: Experiment Protocols

| Protocol | Description |
| --- | --- |
| [**Order Oligo**](https://primerize.stanford.edu/protocol/#IDT) <small>(external)<small> | Order primers for PCR assembly |
| [**Make DNA**](https://primerize.stanford.edu/protocol/#PCR) <small>(external)<small> | PCR (polymerase chain reaction) primer assembly of DNA tempaltes, and purification |
| [**Make RNA**](https://primerize.stanford.edu/protocol/#IVT) <small>(external)<small> | IVT (_in vitro_ transcription) of RNA, and purification |
| ~~**Probe RNA**~~ <small>(external)<small> | CM (chemical mapping) and QC (quality check) |
| ~~**MOHCA**~~ <small>(external)<small> | MOHCA-seq (multiplexed Â•OH cleavage analysis with deep sequencing) proximity mapping |

